zeroize/src/lib.rs

```diff
@@ -795,6 +795,25 @@ unsafe fn volatile_set<T: Copy + Sized>(dst: *mut T, src: T, count: usize) {
     }
 }
 
+/// Zeroizes a flat type/struct. Only zeroizes the values that it owns, and it does not work on
+/// dynamically sized values or trait objects. Do not use this function on a type that already
+/// implements `ZeroizeOnDrop`.
+///
+/// # Safety
+/// - The type must not contain references to outside data or dynamically sized data
+/// - This function can invalidate the type if it is used after this function is called on it. It is
+///   advisable to call this method in `impl Drop`.
```

Comment:
I think it would be good to add something here about requiring that the bit pattern of all zeros must be valid for the type

Reply:
I do not think this requirement is needed for this function. We just need to mention that the function may leave the memory in an invalid state for type `T`.

In other words, I think (but not 100% sure) that the following code is safe:
```rust
#[repr(transparent)]
struct Foo(NonZeroU32);

impl Drop for Foo {
    fn drop(&mut self) {
        unsafe { zeroize_flat_type(self); }
    }
}
```

Reply:
I'm not really clear on if that's allowed or not: https://github.com/rust-lang/unsafe-code-guidelines/issues/394

Reply:
Notably it seems like it would be UB with `drop_in_place` or `ManuallyDrop::drop` which both permit access to the value after running the `Drop` impl

Reply:
Both `drop_in_place` and `ManuallyDrop::drop` are `unsafe`, so it should not be a problem. Per ``ManuallyDrop::drop`` docs:

>However, this “zombie” value should not be exposed to safe code, and this function should not be called more than once. To use a value after it’s been dropped, or drop a value multiple times, can cause Undefined Behavior (depending on what drop does).

For now, we probably should write about _validity_ (with `NonZeroU32` as a counter-example), but mention that zeroized state is allowed to break _safety invariants_ of the type.

Reply:
Sure, that sounds good

Reply:
Okay. Feel free to make any changes